### PR TITLE
feat(418): cancel order 2nd pass

### PR DESCRIPTION
**Pull request by sam-keen**
Updated receipt view for cancel orders.

Screenshot 2023-10-24 at 13 00 28

Edge cases for when an order is omitted or the order and market are omitted are handled like so:

Screenshot 2023-10-18 at 10 31 55

Screenshot 2023-10-18 at 10 32 30

### CODE REVIEW

File: frontend/stores/orders-store.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import { SendMessage } from '../contexts/json-rpc/json-rpc-provider.tsx'
+import { RpcMethods } from '../lib/client-rpc-methods.ts'
+
+export type OrdersStore = {
+  loading: boolean
+  getOrderById: (id: string, request: SendMessage) => Promise<{ order: vegaOrder; lastUpdated: number }>
+}
+
+export const useOrdersStore = create<OrdersStore>((set) => ({
+  loading: false,
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:12]: Need an error state here

File: frontend/stores/orders-store.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  async getOrderById(id: string, request: SendMessage) {
+    try {
+      set({ loading: true })
+      const response = await request(RpcMethods.Fetch, { path: `api/v2/order/${id}` })
+      const order = response.order
+
+      if (!order) {
+        throw new Error(`Order with id ${id} not found`)
+      }
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:21]: set({ error: new Error(`Order with id ${id} not found`) })

File: frontend/stores/orders-store.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  loading: boolean
+  getOrderById: (id: string, request: SendMessage) => Promise<{ order: vegaOrder; lastUpdated: number }>
+}
+
+export const useOrdersStore = create<OrdersStore>((set) => ({
+  loading: false,
+  async getOrderById(id: string, request: SendMessage) {
+    try {
+      set({ loading: true })
+      const response = await request(RpcMethods.Fetch, { path: `api/v2/order/${id}` })
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:16]:
```suggestion
      const response = await request(RpcMethods.Fetch, { path: `api/v2/order/${id}` }, true)
```

File: frontend/stores/orders-store.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        throw new Error(`Order with id ${id} not found`)
+      }
+
+      return {
+        order,
+        lastUpdated: Date.now()
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(`Failed to fetch order: ${error.message}`)
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:29]: set here as well

File: frontend/stores/orders-store.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      }
+      throw new Error('Failed to fetch order')
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:31]: set here as well

File: frontend/stores/orders-store.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export type OrdersStore = {
+  loading: boolean
+  getOrderById: (id: string, request: SendMessage) => Promise<{ order: vegaOrder; lastUpdated: number }>
+}
+
+export const useOrdersStore = create<OrdersStore>((set) => ({
+  loading: false,
+  async getOrderById(id: string, request: SendMessage) {
+    try {
+      set({ loading: true })
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:15]:
```suggestion
      set({ loading: true, error: null })
```

File: frontend/stores/orders-store.spec.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+const mockOrderData = { id: 'some-id', name: 'Test Order' }
+
+describe('Orders Store', () => {
+  const initialOrderState = useOrdersStore.getState()
+  let requestMock: jest.Mock
+
+  beforeAll(() => {
+    jest.useFakeTimers()
+    Date.now = jest.fn(() => 0)
```
Comment (dexturr) [inline frontend/stores/orders-store.spec.ts:12]: jest.setSystemTime rather than mocking out date plz

File: frontend/stores/orders-store.spec.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    requestMock.mockImplementation(
+      () =>
+        new Promise((resolve) => {
+          Promise.resolve().then(() => {
+            resolve({ order: {} })
+          })
+        })
+    )
+
+    await act(async () => {
```
Comment (dexturr) [inline frontend/stores/orders-store.spec.ts:34]: act not needed if we await the promise correctly at the end. See `frontend/stores/assets-store.spec.ts`

File: frontend/components/receipts/utils/order/build-order-columns.tsx [hunk 9]
```diff
@@ -108,5 +131,39 @@ export const buildTypeColumn = (type?: vegaOrderType): [ReactNode, ReactNode] |
 
 export const buildReferenceColumn = (reference?: string): [ReactNode, ReactNode] | null => {
   if (!reference) return null
-  return ['Reference', reference]
+  return [
+    'Reference',
+    <CopyWithCheckmark text={reference} key="order-reference">
+      {truncateMiddle(reference)}
+    </CopyWithCheckmark>
+  ]
```
Comment (dexturr) [inline frontend/components/receipts/utils/order/build-order-columns.tsx:139]: good idea

File: frontend/components/receipts/utils/order/build-order-columns.tsx [hunk 10]
```diff
@@ -8,6 +9,21 @@ import { OrderMarketComponent } from './order-market'
 import { OrderTypeComponent } from './order-type'
 import { PeggedOrderInfo } from './pegged-order-info'
 import { Direction } from './direction'
+import { formatDateWithLocalTimezone } from '@vegaprotocol/utils'
+import { nanoSecondsToMilliseconds } from '../../../../lib/utils.ts'
+import { CopyWithCheckmark } from '../../../copy-with-check'
+
+const orderStatuses = {
```
Comment (dexturr) [inline frontend/components/receipts/utils/order/build-order-columns.tsx:16]: Let's ensure typechecking of all keys:

```suggestion
const orderStatuses: Record<vegaOrderStatus, string> = {
```

File: frontend/lib/transactions.ts [hunk 11]
```diff
@@ -2,6 +2,7 @@ export enum TransactionKeys {
   UNKNOWN = 'unknown',
   ORDER_SUBMISSION = 'orderSubmission',
   ORDER_CANCELLATION = 'orderCancellation',
+  MASS_ORDER_CANCELLATION = 'massOrderCancellation',
```
Comment (dexturr) [inline frontend/lib/transactions.ts:5]: This is currently an accurate representation of possible transaction keys within vega (i.e. the child of the command key in a transaction). I think it's probably better to keep it that way; can we remove this and just put the string into your function above please.

```suggestion
```

File: frontend/lib/transactions.ts [hunk 12]
```diff
@@ -33,6 +34,7 @@ export const TRANSACTION_TITLES: Record<TransactionKeys, string> = {
   [TransactionKeys.UNKNOWN]: 'Unknown',
   [TransactionKeys.ORDER_SUBMISSION]: 'Order Submission',
   [TransactionKeys.ORDER_CANCELLATION]: 'Order Cancellation',
+  [TransactionKeys.MASS_ORDER_CANCELLATION]: 'Mass Order Cancellation',
```
Comment (dexturr) [inline frontend/lib/transactions.ts:37]: See other comment

Comment (dexturr) [inline frontend/lib/transactions.ts:37]: See other comment about transaction keys

```suggestion
```

File: frontend/components/modals/transaction-modal/transaction-header.tsx [hunk 13]
```diff
@@ -10,6 +10,22 @@ export const locators = {
...
+  const type = Object.keys(transaction)[0] as TransactionKeys
+
+  // If we have a mass order cancellation (i.e. no order ID, with or without a market ID)
+  // then we want to display a different title
+  if (
+    type === TransactionKeys.ORDER_CANCELLATION &&
+    transaction.orderCancellation &&
+    !transaction.orderCancellation.orderId
+  ) {
+    return TRANSACTION_TITLES[TransactionKeys.MASS_ORDER_CANCELLATION]
```
Comment (dexturr) [inline frontend/components/modals/transaction-modal/transaction-header.tsx:23]:
```suggestion
    return "Mass Order Cancellation"
```

Comment (dexturr) [reply frontend/components/modals/transaction-modal/transaction-header.tsx:23]: See other comments about removing this for why 😄

File: frontend/components/receipts/orders/cancellation/index.tsx [hunk 14]
```diff
@@ -1,9 +1,65 @@
...
+
+  useEffect(() => {
+    if (orderId) {
+      getOrderById(orderId, request)
+        .then((result) => {
+          setOrderDetails({ order: result.order, lastUpdated: result.lastUpdated })
+        })
+        .catch((err) => {
+          console.error('Failed to fetch order details:', err)
+        })
```
Comment (dexturr) [inline frontend/components/receipts/orders/cancellation/index.tsx:48]: Error handling should be done by the global catch we have. WIth your current setup I think it may be straight up impossible to hit this code branch. Let's kill it.

```suggestion
```

File: frontend/components/receipts/orders/cancellation/index.tsx [hunk 15]
```diff
@@ -1,9 +1,65 @@
...
 
 export const CancellationView = ({ cancellation }: { cancellation: any }) => {
-  return <OrderTable {...cancellation} />
+  const { orderId, marketId } = cancellation
+  const [orderDetails, setOrderDetails] = useState<{ order: any; lastUpdated: number | null }>({
+    order: null,
+    lastUpdated: null
+  })
+
+  const { getOrderById } = useOrdersStore(selectGetOrderById)
```
Comment (dexturr) [inline frontend/components/receipts/orders/cancellation/index.tsx:37]: Once we have an error state, depending on is this goes down first or the error PR (it's lilkely going to be the error PR as it's smaller) we will need to pass that error into `RecieptView` component to ensure error states are shown appropriately.

Suggest making cancellation view presentational and moving all loading logic into the component below.

File: frontend/components/receipts/orders/cancellation/index.tsx [hunk 16]
```diff
@@ -1,9 +1,65 @@
...
+      ) : (
+        <Notification intent={Intent.Warning} message={'Cancel all open orders in all markets'} />
+      )}
+    </div>
+  )
+}
+
+export const selectGetOrderById = (state: OrdersStore) => ({ getOrderById: state.getOrderById })
 
 export const CancellationView = ({ cancellation }: { cancellation: any }) => {
```
Comment (dexturr) [inline frontend/components/receipts/orders/cancellation/index.tsx:30]: Can we move into a different file please as it's not quite complicated

File: frontend/components/receipts/orders/cancellation/index.tsx [hunk 17]
```diff
@@ -1,9 +1,65 @@
...
+  })
+
+  const { getOrderById } = useOrdersStore(selectGetOrderById)
+  const { request } = useJsonRpcClient()
+
+  useEffect(() => {
+    if (orderId) {
+      getOrderById(orderId, request)
+        .then((result) => {
+          setOrderDetails({ order: result.order, lastUpdated: result.lastUpdated })
```
Comment (dexturr) [inline frontend/components/receipts/orders/cancellation/index.tsx:44]: This is incorrect based on Barneys feedback. This should be the datetime that the order was **requested** not the time it was **last updated**. Last updated could be days ago which does not let me know if the data of the order I am seeing might be out of date or not if I have taken actions on another wallet.

Suggest making this a store prop and setting before kicking off the request and nulling out on store destory.

File: examples/sample-dapp/index.html [hunk 18]
```diff
@@ -1,233 +1,228 @@
 <!DOCTYPE html>
```
Comment (dexturr) [inline examples/sample-dapp/index.html:1]: can we revert the changes to this file please, it's just noise and we should leave the example dapp with a transfer transaction as this is deployed public facing (technically).

File: frontend/stores/orders-store.ts [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+      const order = response.order
+
+      if (!order) {
+        throw new Error(`Order with id ${id} not found`)
+      }
+
+      return {
+        order,
+        lastUpdated: Date.now()
+      }
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:26]: Rather than returning let's set this on the store as that is the pattern zustand advocates for and in case we want to split components later this removed the need to prop drill.

Comment (sam-keen) [reply frontend/stores/orders-store.ts:26]: I thought about setting it on the store, but the info is really only pertinent to the specific order that's returned, there's no added benefit for there to be access to the value anywhere else, it just creates more code requiring coverage, ensuring old updatedAt data is ignored etc.

File: frontend/stores/orders-store.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        return
+      }
+
+      return order
+    } catch (error) {
+      if (error instanceof Error) {
+        set({ error: `Failed to fetch order: ${error.message}` })
+      } else {
+        set({ error: 'Failed to fetch order' })
+      }
```
Comment (dexturr) [inline frontend/stores/orders-store.ts:34]:
```suggestion
        set({ error: error as Error })
```

File: frontend/components/receipts/orders/cancellation/cancellation-notification.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+      getMarketById: () => ({
+        id: '1'
+      })
+    })
+    render(<CancellationNotification orderId="" marketId="some-market-id" />)
+
+    expect(screen.getByText('Cancel ALL open orders in some-m…t-id')).toBeInTheDocument()
+  })
+
+  it('should render nothing order id is present', () => {
```
Comment (dalebennett1992) [inline frontend/components/receipts/orders/cancellation/cancellation-notification.spec.tsx:53]: I find this test name a little unclear

**Comment by dalebennett1992** [comment]
test evidence:
For a missing order id, this looks ok.
image

**Error warning misleading when order does not exist**
image

Should we have something specific to the order not being present instead of this

### Error for a market id that does not exist
image

I think we should have more specific handling than bombing out here.

**Review by dalebennett1992 — APPROVED** [review]
lez do dis

**Comment by emilbayes** [comment]
Merged in https://github.com/vegaprotocol/vegawallet-browser/commit/44a8e573bc05174bf4bcd5c5d6f89fb002155e01
